File: firefly/src/main/java/com/firefly/server/http2/router/impl/AbstractPatternMatcher.java
package com.firefly.server.http2.router.impl;


import com.firefly.server.http2.router.Matcher;
import com.firefly.server.http2.router.Router;
import com.firefly.utils.pattern.Pattern;

import java.util.*;

/**
 * @author Pengtao Qiu
 */
abstract public class AbstractPatternMatcher implements Matcher {

    protected Map<PatternRule, Set<Router>> patternMap;

    protected static class PatternRule {
        final String rule;
        final Pattern pattern;

        protected PatternRule(String rule) {
            this.rule = rule;
            pattern = Pattern.compile(rule, "*");
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            PatternRule that = (PatternRule) o;
            return Objects.equals(rule, that.rule);
        }

        @Override
        public int hashCode() {
            return Objects.hash(rule);
        }
    }

    protected Map<PatternRule, Set<Router>> patternMap() {
        if (patternMap == null) {
            patternMap = new HashMap<>();
        }
        return patternMap;
    }

    @Override
    public void add(String rule, Router router) {
        patternMap().computeIfAbsent(new PatternRule(rule), k -> new HashSet<>()).add(router);
    }

    @Override
    public MatchResult match(String v) {
        if (patternMap == null) {
            return null;
        }

        Set<Router> routers = new HashSet<>();
        Map<Router, Map<String, String>> parameters = new HashMap<>();

        patternMap.forEach((rule, routerSet) -> {
            String[] strings = rule.pattern.match(v);
            if (strings != null) {
                routers.addAll(routerSet);
                if (strings.length > 0) {
                    Map<String, String> param = new HashMap<>();
                    for (int i = 0; i < strings.length; i++) {
                        param.put("param" + i, strings[i]);
                    }
                    routerSet.forEach(router -> parameters.put(router, param));
                }
            }
        });
        if (routers.isEmpty()) {
            return null;
        } else {
            return new MatchResult(routers, parameters, getMatchType());
        }
    }
}

File: firefly-net/src/main/kotlin/com/fireflysource/net/http/server/impl/matcher/AbstractPatternMatcher.kt
package com.fireflysource.net.http.server.impl.matcher

import com.fireflysource.common.string.Pattern
import com.fireflysource.net.http.server.Matcher
import com.fireflysource.net.http.server.Router
import java.util.*


abstract class AbstractPatternMatcher : Matcher {

    companion object {
        const val paramName = "param"
    }

    protected val patternMap: MutableMap<PatternRule, SortedSet<Router>> = HashMap<PatternRule, SortedSet<Router>>()

    class PatternRule(val rule: String) {

        val pattern: Pattern = Pattern.compile(rule, "*")

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as PatternRule
            return rule == other.rule
        }

        override fun hashCode(): Int {
            return rule.hashCode()
        }
    }

    override fun add(rule: String, router: Router) {
        patternMap.computeIfAbsent(PatternRule(rule)) { TreeSet() }.add(router)
    }

    override fun match(value: String): Matcher.MatchResult? {
        if (patternMap.isEmpty()) return null

        val routers = TreeSet<Router>()
        val parameters = HashMap<Router, Map<String, String>>()

        patternMap.forEach { (rule, routerSet) ->
            val strings: Array<String>? = rule.pattern.match(value)
            if (strings != null) {
                routers.addAll(routerSet)
                if (strings.isNotEmpty()) {
                    val param: MutableMap<String, String> = HashMap()
                    for (i in strings.indices) {
                        param["$paramName$i"] = strings[i]
                    }
                    routerSet.forEach { router -> parameters[router] = param }
                }
            }
        }
        return if (routers.isEmpty()) null else Matcher.MatchResult(routers, parameters, matchType)
    }
}